Allow for automatic content discovery for cross-mounting blobs (#275)
opencontainers/distribution-spec: spec.md

header returns the canonical digest of the uploaded blob which MAY differ from the provided digest. Most clients MAY
ignore the value but if it is used, the client SHOULD verify the value against the uploaded blob data.

The registry MAY treat the `from` parameter as optional.

> If MAY cross-mount the blob referred to by the `digest` even if the `from` paramer is omitted or the repository in passed in the `from` does not container the blob.

Reviewer: ```suggestion
It MAY cross-mount the blob referred to by the `mount` parameter even if the `from` parameter is omitted or the `from` repository does not contain the blob.
```

Author: Thanks.

Alternatively, if a registry does not support cross-repository mounting or is unable to mount the requested blob,

> it SHOULD return a `202`. This indicates that the upload session has begun and that the client MAY proceed with the upload.

Reviewer 2: Do we need to specify the `202` response will also include a `Location` header for uploading the blob?

Author: I think that's unrelated to this change, right?

header returns the canonical digest of the uploaded blob which MAY differ from the provided digest. Most clients MAY
ignore the value but if it is used, the client SHOULD verify the value against the uploaded blob data.

> The registry MAY treat the `from` parameter as optional, and it MAY cross-mount the blob if it can be found.

Reviewer 3: Should we change `if it can be found` to `if it can be found and can be accessed`?

Reviewer 4: to me, that ACL implication is included in `can be found`, because ideally if it can not be accessed then it should be found. Or is there another use case you're thinking about?

Author: Although it is outside of the specification, I would suggest that we indicate that it is not a good idea to implement this feature on any registry with cross-repo ACLs due to the possibility of information disclosure. I do not think `found` implies any notion of "access" in the sense of "Security"

So, I would rather say something like:
"If the registry does not implement a security model which allows for attenuation of access to reading blobs, it MAY treat the `from` parameter as optional, and it MAY cross-mount the blob if it can be found."

Reviewer 4: 👍

Reviewer 5: Agreed on adding more to wording here. While we don't define an ACL model, it is a potential pitfall with a very small change that is worth calling out. The updated language is reasonable, but I think it can be simplified some to just mentioned the read access. Like @jonjohnsonjr brought it, it could be that the registry only searches in a list of a blob that are known public and therefore able to be read by anyone.

Reviewer 6: We could leverage this for MCR/ACR as well and for avoiding pushing these public well known layers multiple times.
